components/VM_Arm/src/main.c

```diff
@@ -992,10 +992,10 @@ static int alloc_vm_device_cap(uintptr_t addr, vm_t *vm, vm_frame_t *frame_resul
     err = vka_utspace_alloc_at(vm->vka, &frame, kobject_get_type(KOBJECT_FRAME, seL4_PageBits), seL4_PageBits, addr,
                                &cookie);
     if (err) {
-        ZF_LOGE("Grabbing the entire cap for device memory");
+        ZF_LOGV("Grabbing the entire cap for device memory");
         err = simple_get_frame_cap(vm->simple, (void *)addr, seL4_PageBits, &frame);
         if (err) {
-            ZF_LOGE("Failed to grab the entire cap");
+            ZF_LOGV("Failed to grab the entire cap for addr 0x%"PRIxPTR, addr);
```

Comment:
I wonder if the root problem is, that this function should not be used for RAM-backed device install then? Can't this be detected and then this is not tried at all. So the first message can become a `ZF_LOGW()` and this one remains a `ZF_LOGE()`

Reply:
I think the On Demand order of operations is this:

1. Check for untyped object device capability (if you defined the vm0.dtb {})
2. Check for frame (vm0.mmios)
3. Map the address into RAM

That order makes sense to me. Especially since the On Demand Install is called from a memory fault, and afaik there isn't a way to differentiate between whether that access was for RAM, or for a device.